### DISCUSSION TITLE
Shared code generators

### DISCUSSION THREAD

**Discussion by tonai** [vercel/turborepo#6822] [Help] [answered]
Hello,
I have a bunch of projects inside a monorepo, and they all share the same structure.
I wonder if I could have a shared generator that is declared in one place but could be applied to any project I want.
Or should I create a generator at the root, ask for the target package and generate the right path ?
Is there maybe a helper I can use for that ? A way to simply retrieve the modules available in my monorepo ?

**Comment by anthonyshew** [comment]
From what you've described, it sounds to me like this is what you are looking for: https://turbo.build/repo/docs/core-concepts/monorepos/code-generation#copy-an-existing-workspace

Does that seem right?

**Comment by tonai** [comment, marked as answer]
Thanks for your answer.
But not really because copying will increase the number of files and make maintenance more difficult.
For now I have a root generator and a function to list the projects in the monorepo:
```js
async function getPackages() {
  const packages = [];
  const src = './packages';
  const dirs = await readdir(src);
  for (const dir of dirs) {
    const filepath = join(src, dir, 'package.json');
    const file = await readFile(filepath, { encoding: 'utf-8' });
    if (file) {
      try {
        const json = JSON.parse(file);
        if (!json.private) {
          packages.push(dir);
        }
      } catch (error) {
        console.error(error);
      }
    }
  }
  return packages;
}
```
I use it in the prompts to choose the target project:
```js
const { dir } = await inquirer.prompt({
  choices: await getPackages(),
  message: 'For what package ?',
  name: 'dir',
  type: 'list',
});
```

**Reply by anthonyshew** [reply]
Glad you sorted out what you need. Happy building!
